File: C++SrcFiles/FireMageScript.cpp
#include <iostream>
#include <vector>
#include <string>
#include <cmath>
#include <algorithm>

using namespace std;

class FireMageScript : public MonoBehaviour
{
private:
    Animator* animator;
    GameObject* player;
    float playerDistance;
    float chaseSpeed;
    bool entry;
    float travelTime;
    float endEntry;
    bool isFiring;
    public GameObject* attack;
    public Transform firespawn;
    float attackRange;
    public LayerMask targetLayer;
    int maxHealth;
    int currHealth;
    int attackDamage;
    HealthBarScript* healthbar;

public:
    FireMageScript()
    {
        maxHealth = currHealth = 100;
        attackRange = 5.0f;
        attackDamage = 40;
        isFiring = false;
        chaseSpeed = 0.0f;
        animator->SetFloat("Speed", 0.0f); // This will trigger the conditions on which the key animation of this character changes according to the conditions set on the float "Speed".

        player = GameObject::FindGameObjectWithTag("Player");
        healthbar->SetMaxHealth(maxHealth);
        healthbar->SetHealth(currHealth);
    }

    void Update()
    {
        playerDistance = Vector2::Distance(transform->position, player->transform->position);

        // Flip the character if going left;
        Vector3 characterScale = transform->localScale;
        if (player->transform->position.x > transform->position.x && !isFiring) {
            characterScale.x = -18.0f;
        } else if (player->transform->position.x < transform->position.x && !isFiring) {
            characterScale.x = 18.0f;
        }
        transform->localScale = characterScale;

        if (travelTime > 0.0f && entry) {
            chaseSpeed = 4.0f;
            animator->SetFloat("Speed", 4.0f);
            transform->position += (Vector3::left * chaseSpeed) * Time.deltaTime;
            travelTime -= 1.0f * Time.deltaTime;
            if (travelTime <= endEntry) {
                entry = false;
                print("Finished Entering the map!");
            }
        }
        // Movement logic;
        else if (entry == false) {
            // Shoot range for Fire Mage;
            if (playerDistance < 16.0f) {
                if (!isFiring) {
                    Shoot();
                }
            }
            // Run Range for Fire Mage
            else if (playerDistance < 30.0f && playerDistance > 16.0f && !isFiring) {
                Run();
            }
            // Walk Range for Fire Mage
            else if (playerDistance > 30.0f) {
                Walk();
            }
        }

        if (currHealth <= 1) {
            FireMageDeath();
        }
    }

private:
    void FireMageDeath()
    {
        chaseSpeed = 0.0f;
        isFiring = false;
        animator->SetTrigger("Dead");
    }

    void Run()
    {
        chaseSpeed = 14.0f;
        animator->SetFloat("Speed", 10.0f);
        transform->position = Vector2::MoveTowards(transform->position, player->transform->position, chaseSpeed * Time.deltaTime);
    }

    void Walk()
    {
        isFiring = false;
        chaseSpeed = 5.0f;
        animator->SetFloat("Speed", 4.0f);
        transform->position = Vector2::MoveTowards(transform->position, player->transform->position, chaseSpeed * Time.deltaTime);
    }

    void Shoot()
    {
        isFiring = true;
        chaseSpeed = 0.0f;
        animator->SetFloat("Speed", 0.0f);
        FireAttack();
    }

    void FireAttack()
    {
        // Melee attack damage logic;
        if (isFiring) {
            animator->SetTrigger("FireBall");

            // Get all the colliders within the attack range and layer mask
            Collider2D** hitTargets = Physics2D::OverlapCircleAll(firespawn->position, attackRange, targetLayer);

            // Iterate through the hit targets and apply damage
            for (Collider2D* target : *hitTargets) {
                target->GetComponent<SlimeScript>()->TakeDamage(attackDamage);
                print("hit slime!");
            }

            // Free the allocated memory
            delete[] hitTargets;
        }
    }

    void OnDrawGizmosSelected() {
        if (firespawn == nullptr) {
            return;
        }

        Gizmos::DrawWireSphere(firespawn->position, attackRange);
    }

    void TakeDamage(int damage) {
        currHealth -= damage;
        healthbar->SetHealth(currHealth);
    }

}

File: C#SrcFiles/FireMageScript.cs
using System.Collections;
using System.Collections.Generic;
using System.Runtime.CompilerServices;
using Unity.VisualScripting;
using UnityEngine;
using static UnityEngine.EventSystems.EventTrigger;

public class FireMageScript : MonoBehaviour
{
    
    public Animator animator;
    private GameObject player;
    private float PlayerDistance;
    private float chaseSpeed;
    private bool Entry;
    private float traveltime;
    private float EndEntry;
    // to check if the mage is firing at the given moment;
    // fire trigger object;
    public GameObject attack;
    public Transform firespawn;
    private float attackRange;
    public LayerMask TargetLayer;
    // Start is called before the first frame update
    private bool Isfiring;
    public int maxHealth;
    public int currHealth;
    private int attackdamage;
    public HealthBarScript healthbar;

    void Start()
    {
        
        maxHealth = currHealth = 100; ;
        attackRange = 5f;
        attackdamage = 40;
        Isfiring = false;
        chaseSpeed = 0f;
        animator.SetFloat("Speed", 0); // this will trigger the conditions on which the key animation of this character changes according to the conditions set on the float "Speed".

        player = GameObject.FindGameObjectWithTag("Player");
        healthbar.SetMaxHealth(maxHealth);
        healthbar.SetHealth(currHealth);

    }

    // Update is called once per frame
    void Update()
    {
        PlayerDistance = Vector2.Distance(transform.position, player.transform.position);

        // Flip the character if going left;
        Vector3 characterScale = transform.localScale;
        if (player.transform.position.x > transform.position.x && !Isfiring)
        {
            characterScale.x = -18f;
        }
        else if (player.transform.position.x < transform.position.x && !Isfiring)
        {
            characterScale.x = 18f;
        }
        transform.localScale = characterScale;

        
        if (traveltime > 0 && Entry)
        {
            chaseSpeed = 4f;
            animator.SetFloat("Speed", 4f);
            transform.position += (Vector3.left * chaseSpeed) * Time.deltaTime;
            traveltime -= 1 * Time.deltaTime;
            if (traveltime <= EndEntry)
            {
                Entry = false;
                print("Finished Entering the map!");
            }
        }
        // movement logic;
        else if (Entry == false) {

            
            // Shoot range for Fire Mage;
            if (PlayerDistance < 16f)
            {
                if(!Isfiring)
                {
                    Shoot();
                }

            }
            // Run Range for Fire Mage
            else if (PlayerDistance < 30f && PlayerDistance > 16f && !Isfiring)
            {
                Run();
            }
            // Walk Range for Fire Mage
            else if (PlayerDistance > 30f)
            {
                Walk();
            }   
        }

        if (currHealth <= 1)
        {
            FireMageDeath();
        }


    }

    

    private void FireMageDeath()
    {
        chaseSpeed = 0;
        Isfiring = false;
        animator.SetTrigger("Dead");
    }

    private void Run()
    {
        chaseSpeed = 14f;
        animator.SetFloat("Speed", 10f);
        transform.position = Vector2.MoveTowards(this.transform.position, player.transform.position, chaseSpeed * Time.deltaTime);
    }

    private void Walk()
    {
        Isfiring = false;
        chaseSpeed = 5f;
        animator.SetFloat("Speed", 4f);
        transform.position = Vector2.MoveTowards(this.transform.position, player.transform.position, chaseSpeed * Time.deltaTime);
    }

    private void Shoot()
    {
        Isfiring = true;
        chaseSpeed = 0f;
        animator.SetFloat("Speed", 0);
        FireAttack();
    }

    private void FireAttack()
    {
        // Melee attack damage logic;
        if (Isfiring)
        {
            animator.SetTrigger("FireBall");
            Collider2D[] HitTarget = Physics2D.OverlapCircleAll(firespawn.position, attackRange, TargetLayer);
            foreach (Collider2D target in HitTarget)
            {
                // giving the slime 50 damage for the fire attack;
                target.GetComponent<SlimeScript>().TakeDamage(attackdamage);
                print("hit slime!");
            }
        }
    }

    void OnDrawGizmosSelected()
    {
        if (firespawn == null)
        {
            return;
        }
        Gizmos.DrawWireSphere(firespawn.position, attackRange);
    }
    public void TakeDamage(int Damage)
    {
        currHealth -= Damage;
        healthbar.SetHealth(currHealth);
    }
}
